lib/master/master.go

```diff
@@ -146,8 +146,8 @@ func (m *Master) Run() {
 		case beat := <-m.heartbeatChan:
 			// update task statuses
 			// check whether a job has completed all its tasks
-			m.instances[beat.addr].heartbeatChan <- true
 			m.updateOnHeartbeat(beat)
+			m.instances[beat.addr].heartbeatChan <- true
```

Comment:
It's not a deadlock I don't think, the code is just stuck there for some reason. Well, the reason is that we create a new slave in `updateOnHeartbeat`, but that function is originally called after we write to `m.instances[beat.addr].heartbeatChan`. Effectively we're writing to a channel that does not exist yet. Not sure why the code is stuck there, but it works now.